### PR TITLE
Introduce lockfile for monthly benchs

### PR DESCRIPTION
**Pull request by maiste**

### CODE REVIEW

File: lockfiles/irmin-22-02.opam.locked [hunk 1]
```diff
@@ -0,0 +1,1914 @@
...
+  "result" {= "1.5"}
+  "rresult" {= "0.6.0+dune"}
+  "rusage" {= "1.0.0"}
+  "semaphore-compat" {= "1.0.1"}
+  "seq" {= "base+dune"}
+  "sexplib" {= "v0.14.0"}
+  "sexplib0" {= "v0.14.0"}
+  "stdlib-shims" {= "0.3.0"}
+  "stringext" {= "1.6.0"}
+  "tcpip" {= "7.0.1"}
```
Comment (samoht) [inline lockfiles/irmin-22-02.opam.locked:178]: any idea why this is here? /cc @Leonidas-from-XIV

**Comment by samoht** [comment]
How did you generate the lockfiles? There are a lot of unnecessary dependencies (for instance `tcpip` and all the `git` related packages). Would be nice to just pull in only what is needed to build and run the benchmarks to minimize churn if possible.

Note: it's not super critical so I'm fine to merge those as a first step and fix this later.

**Comment by maiste** [comment]
I pinned the version of `index` and `repr` to the right month, added `opam-overlays` to the `opam repo` and then run `opam monorepo lock`. AFAIU, `monorepo` tries to figure out all the dependencies from all the packages to make sure to have a self-contained "package space". As we have `irmin-mirage` and `irmin-git`, I think it tries to solve these dependencies too. Of course, we can specify the packages we need to run benchmarks but IMHO, keeping as much information as possible is important. If we want to go back in time for any reason, we won't struggle with dependencies. Moreover, it might increase the amount of work required to make it run. As it doesn't take too much space and too much time to download, it doesn't seem to be a bad trade off. 

 @samoht, what do you think? :thinking: 

PS: this is only a draft PR for now, I want to have the maximum of lock files before merging :smile:

**Comment by samoht** [comment]
Indeed it's better to be able to rebuild the full tree if needed. Let's keep it that way then.

I guess the next step is to update the script to use the lockfile :-)

**Comment by maiste** [comment]
Ok!

Indeed, the plan is first to update the point in time with vendors directory to not need this vendoring (as you suggested offline) and only rely on the lock file.
Then I'll update the script but it shouldn't be hard as it's less complicated :)

File: lockfiles/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+## Description
+
+The goal of these files is to be able to have a reproductible build of `irmin`
+for a benchmark purpose. It is used with `opam monorepo` and `dune`.
+
+## Disclamer
+
+Files older than september 21 only work in the context of `irmin-bench`. They
+are intended to be used with the following configuration:
+  - `ocaml.0.13.x`
```
Comment (samoht) [inline lockfiles/README.md:10]: That seems odd? 4.13.x?

File: lockfiles/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  - `opam-monorepo.0.2.7`
+  - `dune.2.9.3`
+
+:warning: We can't predict the behaviour if it is used with another configuration!
```
Comment (samoht) [inline lockfiles/README.md:14]: I'm not totally sure to understand this warning. Why don't you set these versions in the lock files?

File: lockfiles/README.md (hunk 2 above)
Comment (maiste) [reply lockfiles/README.md:10]: It works with ocaml `4.13` and AFAIU, the other versions are here to correct a bug with `gcc` (4.13.1 and 4.13.2).

File: lockfiles/README.md (hunk 3 above)
Comment (maiste) [reply lockfiles/README.md:14]: There are two issues:
 * `opam monorepo` is needed to make `opam monorepo` work of course. As we need `dune` to build the code and `ocaml` to run it, even if we set the version in the lock file, it won't install it in `duniverse` (or we need to use it with  `opam install [..] --locked` but it doesn't take advantage of `monorepo`).
 * Some errors are raised when you are trying to compile `irmin` (monorepo version) with `dune.3.0.x` as it removes the builtin of `uchar`. If we want it to be compatible it requires to update the code for each branch. As we wanted to have the benchmarks with lock files quickly, we have decided to just pin the `dune` dependency.

Does it seem ok for you?

File: lockfiles/README.md (hunk 2 above)
Comment (maiste) [reply lockfiles/README.md:10]: My mistake, it's a typo! I'll correct it in the next PR :+1: 
(Here: 1ec5ca0ba3742ad5821592c10e102a7dc9aec082)

**Pull request merged**